pkg/component/runtime/manager.go

```diff
@@ -646,40 +649,81 @@ func (m *Manager) update(components []component.Component, teardown bool) error
 	}
 
 	touched := make(map[string]bool)
+	newComponents := make([]component.Component, 0, len(components))
 	for _, comp := range components {
 		touched[comp.ID] = true
-		existing, ok := m.current[comp.ID]
-		if ok {
+		if existing, ok := m.current[comp.ID]; ok {
 			// existing component; send runtime updated value
 			existing.currComp = comp
 			if err := existing.runtime.Update(comp); err != nil {
 				return fmt.Errorf("failed to update component %s: %w", comp.ID, err)
 			}
-		} else {
-			// new component; create its runtime
-			logger := m.baseLogger.Named(fmt.Sprintf("component.runtime.%s", comp.ID))
-			state, err := newComponentRuntimeState(m, logger, m.monitor, comp)
-			if err != nil {
-				return fmt.Errorf("failed to create new component %s: %w", comp.ID, err)
-			}
-			m.current[comp.ID] = state
-			err = state.start()
-			if err != nil {
-				return fmt.Errorf("failed to start component %s: %w", comp.ID, err)
-			}
+			continue
 		}
+		newComponents = append(newComponents, comp)
 	}
+
 	for id, existing := range m.current {
 		// skip if already touched (meaning it still existing)
 		if _, done := touched[id]; done {
 			continue
 		}
 		// component was removed (time to clean it up)
 		_ = existing.stop(teardown)
+		// stop is async, wait for operation to finish,
+		// otherwise new instance may be started and components
+		// may fight for resources (e.g ports, files, locks)
+		m.waitForStopped(existing)
```

Comment:
Nitpicking here: can't we stop and wait for stopped services in parallel using goroutines and a WaitGroup? This would allow us to be a bit faster instead of shutting down one component at a time... Maybe this can be a further optimization down the line ?

Reply:
Yes this should absolutely happen in parallel.